File: backend/app/models/ticket.py
from enum import Enum
from datetime import datetime, timezone
from bson import ObjectId
from pydantic import BaseModel, Field


class PyObjectId(ObjectId):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return ObjectId(v)

    @classmethod
    def __get_pydantic_json_schema__(cls, schema, handler):
        schema.update(type="string")
        return schema


class Priority(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class TicketStatus(str, Enum):
    open = "open"
    resolved = "resolved"


class TicketDocument(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    title: str
    description: str
    priority: Priority
    status: TicketStatus = TicketStatus.open
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    resolved_at: datetime | None = None

    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_encoders": {ObjectId: str},
    }

File: frontend/src/types/ticket.ts
export type Priority = 'low' | 'medium' | 'high';
export type TicketStatus = 'open' | 'resolved';

export interface Ticket {
  id: string;
  title: string;
  description: string;
  priority: Priority;
  status: TicketStatus;
  created_at: string;
  resolved_at: string | null;
}

export interface CreateTicketPayload {
  title: string;
  description: string;
  priority: Priority;
}

export interface TicketListResponse {
  tickets: Ticket[];
  total: number;
}
